go/vt/vtgate/vcursor_impl.go

```diff
@@ -929,6 +930,31 @@ func (vc *vcursorImpl) SetPriority(priority string) {
 	}
 }
 
+func (vc *vcursorImpl) SetExecQueryTimeout(timeout *int) {
+	// Determine the effective timeout: use passed timeout if non-nil, otherwise use session's query timeout if available
+	var execTimeout *int
+	if timeout != nil {
+		execTimeout = timeout
+	} else if sessionTimeout := vc.GetQueryTimeout(0); sessionTimeout > 0 {
+		execTimeout = &sessionTimeout
+	}
+
+	// If no effective timeout and no session options, return early
+	if execTimeout == nil {
+		if vc.safeSession.GetOptions() == nil {
+			return
+		}
+		vc.safeSession.GetOrCreateOptions().Timeout = nil
+		return
+	}
+
+	vc.queryTimeout = time.Duration(*execTimeout) * time.Millisecond
+	// Set the authoritative timeout using the determined execTimeout
+	vc.safeSession.GetOrCreateOptions().Timeout = &querypb.ExecuteOptions_AuthoritativeTimeout{
+		AuthoritativeTimeout: int64(*execTimeout),
+	}
```

Comment:
(continued from above suggestion)
```suggestion

	vc.queryTimeout = time.Duration(execTimeout) * time.Millisecond
	// Set the authoritative timeout using the determined execTimeout
	vc.safeSession.GetOrCreateOptions().Timeout = &querypb.ExecuteOptions_AuthoritativeTimeout{
		AuthoritativeTimeout: int64(execTimeout),
	}
```